chainer/training/trainer.py

```diff
@@ -0,0 +1,285 @@
+      training status.
+
+    The current state of the trainer object and objects handled by the trainer
+    can be serialized through the standard serialization protocol of Chainer.
+    It enables us to easily suspend and resume the training loop.
+
+    .. note::
+       The serialization does not recover everything of the training loop. It
+       only recovers the states which change over the training (e.g.
+       parameters, optimizer states, the batch iterator state, extension
+       states, etc.). You must initialize the objects correctly before
+       deserializing the states.
+
+       On the other hand, it means that users can change the settings on
+       deserialization. For example, the exit condition can be changed on the
+       deserialization, so users can train the model for some iterations,
+       suspend it, and then resume it with larger number of total iterations.
+
+    During the training, it also creates a :class:`~chainer.Reporter` object to
+    store observed values on each update. For each iteration, it creates a
+    fresh observation dictionary and stores it in the :attr:`observation`
+    attribute.
+
+    Links of the target model of each optimizer are registered to the reporter
+    object as observers, where the name of each observer is constructed as the
+    format ``<optimizer name><link name>``. The link name is given by the
+    :meth:`chainer.Link.namedlink` method, which represetns the path to each
+    link in the hierarchy. Other observers can be registered by accessing the
+    reporter object via the :attr:`reporter` attribute.
+
+    The default trainer is `plain`, i.e., it does not contain any extensions.
+
+    Args:
+        updater (Updater): Updater object. It defines how to update the models.
+        stop_trigger: Trigger that determines when to stop the training loop.
+            If it is not callable, it is passed to :class:`IntervalTrigger`.
+
+    Attributes:
+        updater: The updater object for this trainer.
+        stop_trigger: Trigger that determines when to stop the training loop.
+            The training loop stops at the iteration on which this trigger
+            returns True.
+        observation: Observation of values made at the last update. See the
+            :class:`Reporter` class for details.
+        out: Output directory.
+        reporter: Reporter object to report observed values.
+
+    """
+    def __init__(self, updater, stop_trigger=None, out='result'):
+        self.updater = updater
+        self.stop_trigger = trigger_module.get_trigger(stop_trigger)
+        self.observation = {}
+        self.out = out
+
+        reporter = reporter_module.Reporter()
+        for name, optimizer in six.iteritems(updater.get_all_optimizers()):
+            reporter.add_observer(name, optimizer.target)
+            reporter.add_observers(
+                name, optimizer.target.namedlinks(skipself=True))
```

Comment:
Amm, this is because I wanted to avoid an odd name like "main/" for the target link itself, which is the root of the link hierarchy and has the name '/'.